### PR TITLE
[ntuple] Add support for truncated single-precision float columns

### PR DESCRIPTION
**Pull request by silverweed**
# This Pull request:
adds a new ColumnType `Real32Trunc`, that stores a real value as a tightly-packed, IEEE-754 single precision float using less than 32 bits. The missing bits are simply truncated from the mantissa, causing the value to be rounded towards zero.
The valid range of bit widths is `[10, 31]` inclusive.

This is the first ColumnType with a variable bit width, therefore it requires some extra handling on the RColumnElement and RField side, but it uses a single type id on disk (`0x1D`) and a single enum value `EColumnType::kReal32Trunc`.

The way to use it is by calling the new `SetTruncated(nBits)` method on `RField<float>`.

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

### CODE REVIEW

File: tree/ntuple/v7/src/RColumn.cxx [hunk 1]
```diff
@@ -27,10 +27,9 @@ ROOT::Experimental::Internal::RColumn::RColumn(EColumnType type, std::uint32_t c
                                                std::uint16_t representationIndex)
    : fType(type), fIndex(columnIndex), fRepresentationIndex(representationIndex), fTeam({this})
 {
-   // TODO(jblomer): fix for column types with configurable bit length once available
    const auto [minBits, maxBits] = RColumnElementBase::GetValidBitRange(type);
-   assert(minBits == maxBits);
-   fBitsOnStorage = minBits;
+   (void)minBits;
```
Comment (silverweed) [inline tree/ntuple/v7/src/RColumn.cxx:31]: @jblomer the way I implemented the variable bit length seems to not need this information here, but maybe I'm missing something and/or there is a better way to implement it that you were foreseeing when writing this

File: tree/ntuple/v7/test/ntuple_packing.cxx [hunk 2]
```diff
@@ -5,20 +5,23 @@
 #include <cstring> // for memcmp
 #include <limits>
 #include <type_traits>
+#include "ROOT/RColumnElement.hxx"
 #include <utility>
 
-template <typename PodT, typename NarrowT, ROOT::Experimental::EColumnType ColumnT>
+using ROOT::Experimental::Internal::RColumnElement;
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_packing.cxx:11]: as commented in another PR, putting this in the middle of a 100-line commit makes it very hard to understand which parts were actually changed...

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx [hunk 3]
```diff
@@ -315,7 +315,7 @@ public:
 
    std::size_t GetSize() const { return fSize; }
    std::size_t GetBitsOnStorage() const { return fBitsOnStorage; }
-   std::size_t GetPackedSize(std::size_t nElements = 1U) const { return (nElements * fBitsOnStorage + 7) / 8; }
+   virtual std::size_t GetPackedSize(std::size_t nElements = 1U) const { return (nElements * fBitsOnStorage + 7) / 8; }
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:318]: I think this inadvertently introduces a `virtual` function call in the hot code path of reading and writing simple column types, something that we really try to avoid. If the return cannot be computed anymore (I didn't check the logic), I think it would be better to introduce a new member field that holds the value directly.

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx [hunk 4]
```diff
@@ -692,14 +692,14 @@ public:
 /// Tightly packs `count` floats contained in `src` into `dst` using `nFloatBits` per float.
 /// `nFloatBits` must be >= kReal32TruncBitsMin and <= kReal32TruncBitsMax.
 /// The extra bits are dropped from the mantissa. The sign and exponent bits are always preserved.
-/// IMPORTANT: the size of `dst` must be rounded up from `count * nFloatBits`
-/// to the next multiple of 4 (i.e. the word size).
+/// IMPORTANT: the size of `dst` must be at least `(count * nFloatBits + 63) / 64` (i.e.
+/// it needs to be rounded up to the next word).
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RColumnElement.hxx]: Is this a strict requirement or could the contents also be packed at byte granularity?

**Review by hahnjo — COMMENTED** [review]
I left some high-level comments, but in general the PR is really hard to read and review because there are unrelated changes (both `using namespace` and formatting fixes) packed in the individual commits. The first commit also still has traces of using multiple column types (`kReal32TruncBegin`) that get subsequently removed, which makes it really hard to follow unless reviewing the entire +525 -100 change in one go. If possible, it would be really appreciated to help reviewers reading through the code...

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx (hunk 4 above)
Comment (hahnjo) [reply tree/ntuple/v7/inc/ROOT/RColumnElement.hxx]: Do you have measured evidence that it's actually faster? If all writes go to cache, it might not matter at all. Also, this should only be important for the last word which requires special handling anyway, right? There's also another problem that many platforms (ARM, PPC) require alignment for 4 and 8 byte reads, which might be hard to guarantee if we ever `mmap` pages into memory.

Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RColumnElement.hxx]: Turns out that rewriting the packing code to byte granularity also significantly simplified it :)

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx [hunk 5]
```diff
@@ -754,12 +745,22 @@ public:
 
    void Pack(void *dst, void *src, std::size_t count) const final
    {
-      BitPacking::PackBits(dst, reinterpret_cast<const float *>(src), count, fBitsOnStorage);
+#if R__LITTLE_ENDIAN == 0
+      auto bswapped = std::make_unique<std::uint32_t[]>(count);
+      CopyBswap<sizeof(std::uint32_t)>(bswapped.get(), src, count);
```
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:750]: maybe this could be done in place? Not sure if we are allowed to modify `src` here...

File: tree/ntuple/v7/test/ntuple_packing.cxx [hunk 6]
```diff
@@ -353,10 +360,147 @@ TEST(Packing, OnDiskEncoding)
...
+
    auto reader = RNTupleReader::Open("ntuple", fileGuard.GetPath());
    EXPECT_EQ(EColumnType::kIndex64, reader->GetModel().GetField("str").GetColumnRepresentatives()[0][0]);
    EXPECT_EQ(2u, reader->GetNEntries());
    auto viewStr = reader->GetView<std::string>("str");
    EXPECT_EQ(std::string("abc"), viewStr(0));
    EXPECT_EQ(std::string("de"), viewStr(1));
+   auto viewFtrunc = reader->GetView<float>("float32Trunc");
+   EXPECT_EQ(-3.5f, viewFtrunc(0));
+   EXPECT_EQ(1.75f, viewFtrunc(1));
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_packing.cxx:376]:
```suggestion
   EXPECT_FLOAT_EQ(-3.5f, viewFtrunc(0));
   EXPECT_FLOAT_EQ(1.75f, viewFtrunc(1));
```

**Review by jblomer — CHANGES_REQUESTED** [review]
Very nice!

I think we should add a basic test in `ntuple_endian` for the new type, along the lines of the other tests.

For the testing in packing, I'd suggest in addition one test that is exhaustive and tests in a double for loop all valid bit widths and number of elements from 1 to 2000 or so with random floats, checking that packing and unpacking yield the same result. (Of course for the low bit width, the precision tolerance may need to be adjusted).

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx [hunk 7]
```diff
@@ -683,6 +698,74 @@ public:
...
+
+}; // namespace BitPacking
+
+template <>
+class RColumnElement<float, EColumnType::kReal32Trunc> : public RColumnElementBase {
+public:
+   static constexpr bool kIsMappable = false;
+   static constexpr std::size_t kSize = sizeof(float);
+
+   RColumnElement() : RColumnElementBase(kSize, kReal32TruncBitsMax) {}
```
[the remark this answers is not in the document]
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:734]: I'm not a fan of the idea of defaulting to arbitrary hard-to-predict values. Sure, this is also the case for the current choice, but maybe the proper solution - to ensure that the user always specifies the amount of bit they want - is to set this to an invalid value (0?) and give a proper error message if the user forgets to call `SetBitsOnStorage` (which should not be that common, given that the way a user is expected to generate Real32Trunc values is by calling `SetTruncated` on an RField).

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx [hunk 8]
```diff
@@ -683,6 +698,74 @@ public:
...
 
+namespace BitPacking {
+
+using Word_t = std::uintmax_t;
+constexpr std::size_t kBitsPerWord = sizeof(Word_t) * 8;
+
+/// Returns the minimum safe size (in bytes) of a buffer that is intended to be used as a destination for PackBits
+/// or a source for UnpackBits.
+/// Passing a buffer that's less than this size will cause invalid memory reads and writes.
+constexpr std::size_t MinBufSize(std::size_t count, std::size_t nDstBits)
```
[the remark this answers is not in the document]
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:709]: It's supposed to also be used by an "external user" of the Pack/UnpackBits functions. In our case we don't need it though, because this happens to be the same as `GetBitsOnStorage()` (well, it's no coincidence of course - it was made this way intentionally). I can probably remove it and just say directly in the functions' comments that the proper size is `(count * nBits + 7) / 8`.

File: tree/ntuple/v7/src/RColumnElement.cxx [hunk 9]
```diff
@@ -167,3 +171,126 @@ void ROOT::Experimental::Internal::RColumnElement<bool, ROOT::Experimental::ECol
...
+   R__ASSERT(0 < nSrcBits && nSrcBits <= sizeofDst * 8);
+
+   unsigned char *dstArray = reinterpret_cast<unsigned char *>(dst);
+   const Word_t *srcArray = reinterpret_cast<const Word_t *>(src);
+   const auto nWordsToLoad = (count * nSrcBits + kBitsPerWord - 1) / kBitsPerWord;
+
+   // bit offset of the next packed item inside the currently loaded word
+   int offInWord = 0;
+   std::size_t dstIdx = 0;
+   std::uint32_t prevWordAccum = 0;
```
[the remark this answers is not in the document]
Comment (silverweed) [reply tree/ntuple/v7/src/RColumnElement.cxx:244]: Right, I missed to replace the type somewhere along the lines when I introduced `Word_t`.

As for the name, yes, it's a different kind of "accumulator" and might be better named `prevWordLsb` or similar.

File: tree/ntuple/v7/src/RColumnElement.cxx [hunk 10]
```diff
@@ -167,3 +171,126 @@ void ROOT::Experimental::Internal::RColumnElement<bool, ROOT::Experimental::ECol
...
+
+      // If `offInWord` is negative, it means that the last item was split
+      // across 2 words and we need to recombine it.
+      if (offInWord < 0) {
+         std::size_t nMsb = nSrcBits + offInWord;
+         std::uint32_t msb = packedBytes << (8 * sizeofDst - nMsb);
+         std::uint32_t packedWord = msb | prevWordAccum;
+         prevWordAccum = 0;
+         memcpy(dstArray + dstIdx * sizeofDst, &packedWord, sizeofDst);
+         ByteSwapIfNecessary(dst[dstIdx]);
```
[the remark this answers is not in the document]
Comment (silverweed) [reply tree/ntuple/v7/src/RColumnElement.cxx:265]: Another oversight from refactoring (originally I wrote the pack/unpack to also perform byte swapping), thanks!

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx [hunk 11]
```diff
@@ -683,6 +698,74 @@ public:
...
+   }
+
+   bool IsMappable() const final { return kIsMappable; }
+
+   void Pack(void *dst, void *src, std::size_t count) const final
+   {
+      R__ASSERT(GetPackedSize(count) == BitPacking::MinBufSize(count, fBitsOnStorage));
+
+#if R__LITTLE_ENDIAN == 0
+      auto bswapped = std::make_unique<float[]>(count);
```
[the remark this answers is not in the document]
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:749]: I originally had the byte swap inside the Un/PackBits functions, then I extracted it because I wasn't sure if we would only want to use those functions with 32 bit values.
I guess for now it's the case, so it's probably worth doing the change since it avoids an allocation.

File: tree/ntuple/v7/test/ntuple_packing.cxx (hunk 6 above)
Comment (silverweed) [reply tree/ntuple/v7/test/ntuple_packing.cxx:376]: In this case I specifically want EQ because I know that that number has an exact representation and it should be exactly that bit pattern.

File: tree/ntuple/v7/src/RField.cxx [hunk 12]
```diff
@@ -1387,10 +1388,49 @@ const ROOT::Experimental::RFieldBase::RColumnRepresentations &
...
+         column->SetBitsOnStorage(coldesc.GetBitsOnStorage());
+      }
+      fColumnRepresentatives.emplace_back(onDiskTypes);
+      if (representationIndex > 0) {
+         fAvailableColumns[0]->MergeTeams(*fAvailableColumns[representationIndex]);
+      }
+
+      representationIndex++;
+   } while (true);
+   fPrincipalColumn = fAvailableColumns[0].get();
```
Comment (jblomer) [inline tree/ntuple/v7/src/RField.cxx:1431]: We need to check for the case that (by accident) the field has no attached column at all (throw exception).

File: tree/ntuple/v7/src/RColumnElement.cxx [hunk 13]
```diff
@@ -131,6 +133,7 @@ ROOT::Experimental::Internal::RColumnElementBase::Generate<void>(EColumnType typ
    case EColumnType::kSplitUInt32: return std::make_unique<RColumnElement<std::uint32_t, EColumnType::kSplitUInt32>>();
    case EColumnType::kSplitInt16: return std::make_unique<RColumnElement<std::int16_t, EColumnType::kSplitInt16>>();
    case EColumnType::kSplitUInt16: return std::make_unique<RColumnElement<std::uint16_t, EColumnType::kSplitUInt16>>();
+   case EColumnType::kReal32Trunc: return std::make_unique<RColumnElement<std::uint16_t, EColumnType::kReal32Trunc>>();
```
Comment (jblomer) [inline tree/ntuple/v7/src/RColumnElement.cxx:136]:
```suggestion
   case EColumnType::kReal32Trunc: return std::make_unique<RColumnElement<float, EColumnType::kReal32Trunc>>();
```

**Review by jblomer — CHANGES_REQUESTED** [review]

File: tree/ntuple/v7/src/RField.cxx (hunk 12 above)
Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:1431]: Please disregard, already checked by `EnsureCompatibleColumnTypes`

**Review by jblomer — APPROVED** [review]
Great work!
